### PR TITLE
Allow use in same-origin children, add Feature Policy integration

### LINKED ISSUE #10: Allow use from within secure context and top-level browsing context only
(This issue is branched from a proposal made in https://github.com/w3c/battery/issues/5#issuecomment-257554180 since there seemed to be adequate support for making such a spec update.)

**Problem:**

Malicious content such as framed tracker scripts using the API to fingerprint users.

**Proposed solution:**

Make the API available only within a [secure context](https://w3c.github.io/webappsec-secure-contexts/#secure-context) that is also a [top-level browsing context](https://www.w3.org/TR/html51/browsers.html#top-level-browsing-context). This disallows the use of the API within framed content, as well as from any content that is not a secure context.

See [top-level documents](https://w3c.github.io/webappsec-secure-contexts/#examples-top-level) and [framed documents](https://w3c.github.io/webappsec-secure-contexts/#examples-framed) for illustrations.

**Summary of changes:**

There exists a hook in the spec to implement this change with no API surface changes in a backwards compatible manner:
* if the API is invoked from within a browsing context that is not a secure context and not a top-level browsing context, then
* leave the promise returned by [`getBattery()`](https://w3c.github.io/battery/#h-the-navigator-interface) in a pending state

This means we won't break existing web content using the API.

@riju volunteered to look into updating the Chromium/Blink implementation accordingly after the spec changes have landed.

### PR DESCRIPTION
**Pull request by anssiko**
Per @clelland's suggestion in https://github.com/w3c/battery/issues/10#issuecomment-324661401 this change allows Battery Status API use in same-origin children, and adds Feature Policy integration.

PTAL @clelland @RByers @mounirlamouri @riju

Fixes #10.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/battery/pull/13.html" title="Last updated on Sep 24, 2019, 12:29 PM UTC (6d2ce77)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/battery/13/ce5317b...6d2ce77.html" title="Last updated on Sep 24, 2019, 12:29 PM UTC (6d2ce77)">Diff</a>

### CODE REVIEW

**Comment by anssiko** [comment]
@clelland @RByers @mounirlamouri, any concerns?

If I don't hear any, I'll merge this and @riju will update the Chrome's implementation accordingly.

**Comment by RByers** [comment]
Looks reasonable to me.  Is the plan to add a [feature control hook](https://github.com/WICG/feature-policy/blob/gh-pages/integration.md) later?

**Comment by clelland** [comment]
The only forwards-incompatibility I can see with Feature Policy here is that if a deeply nested frame happens to be same-origin with the top-document, then it will be granted access to the API by default, while feature policy would disallow that.

That is, in an A->B->A embedding scenario, both the top-level and the innermost frame would be able to use the battery API. With feature policy, the inner frame wouldn't be allowed to use it, unless the top-level doc granted it to the middle frame, which then further granted it to the inner one.

(That is, feature policy would allow access to same-origin children -- and their same-origin children, and so on -- but not to arbitrary same-origin descendent frames)

**Comment by mounirlamouri** [comment]
Do we have any idea of how common it is to have x-origin iframe usage of the Battery API? Said otherwise, are we sure we are not breaking a real use case? Are we sure this change is even web compatible?

**Comment by anssiko** [comment]
Thanks for your comments!

>Is the plan to add a feature control hook later?

@RByers, that was my initial plan. That said, we could consider adding the hook on the same pass, there's just one procedural issue that need to be cleared:

The [Feature Policy spec](https://wicg.github.io/feature-policy/) is currently in incubation, and having an unstable normative reference might be an issue if we want to advance the Battery Status API to Candidate Recommendation and beyond.

@dontcallmedom to confirm whether we could make an exception here.

>Do we have any idea of how common it is to have x-origin iframe usage of the Battery API?

@mounirlamouri, I'd like to know as well, but I don't have any data.

If we care strongly about this, would Chrome be fine with us adding telemetry to figure out how common x-origin iframe usage is?

> A->B->A embedding scenario

@clelland, I'd expect many other platform features be similarly impacted? If so, I wouldn't be concerned about this case at hand specifically.

Couple of questions to @clelland et al.:

Feature Policy "v1" shipped in Chrome M62. Does it have all the features implemented we'd need for this spec to integrate with it?

Do you have established conventions for spec authors on how to hook into the Feature Policy spec the right way. I'm looking for normative spec language I could borrow similarly to e.g. https://w3c.github.io/webappsec-secure-contexts/#new This helps keep specs consistent.

Thanks!

**Comment by dontcallmedom** [comment]
re CR transition with a reference to an unstable spec - this won't block transition to CR (although we would want to highlight it in the transition request), but may block transition to PR. That being said, transition to PR needs a 2nd implementation we don't have yet.

**Comment by anssiko** [comment]
Thanks for the confirmation @dontcallmedom.

With that, I'd be happy to hook this spec into the Feature Policy spec when shown the preferred conventions in doing so.

**Comment by clelland** [comment]
@anssiko -- I think that we do have all of the features implemented in Chromium to do that. As long as the plan is just to reject the promise when disallowed, then it should be pretty simple.

I've written up a spec integration guide at https://github.com/WICG/feature-policy/blob/gh-pages/integration.md -- it's not strictly normative, but take a look and see if gives you enough guidance for the wording. If there's anything that I can add to that to help you out, I'm glad to do it.

**Comment by anssiko** [comment]
@clelland, thanks, the integration guide is very helpful.

I added Feature Policy integration in https://github.com/w3c/battery/pull/13/commits/c2083ad3126b629cae0f5326e37033f25b217ed1, PTAL. See also the [preview](https://rawgit.com/w3c/battery/issue-10-secure-top-level/index.html#feature-policy-integration).

(I assumed the integration guide means `"SecurityError" DOMException`.)

**Comment by clelland** [comment]
@anssiko: Yes, that should totally read `SecurityError DOMException`, thanks :) PR fired. The actual error, though, should be whatever makes sense for the API -- whether rejecting a promise, or returning an error code, the important thing is just defining some way for the API to fail when disabled.

This change LGTM, though -- thanks!

**Comment by anssiko** [comment]
@mounirlamouri PTAL and merge this PR if you're happy.

**Comment by anssiko** [comment]
@mounirlamouri, gentle ping.

Others on this PR seem to be fine with this change, would like to get your explicit OK prior to merging.

**Comment by anssiko** [comment]
@mounirlamouri, we'll merge this PR unless we hear concerns from you by next Tuesday 24 Oct, to be followed by a corresponding patch to Chromium.

**Comment by anssiko** [comment]
@Honry, as we prepare for this PR to land, could you evaluate the gaps in the [battery-status test suite](https://github.com/w3c/web-platform-tests/tree/master/battery-status) (in the spirit of [test as you commit](https://w3c.github.io/dap-charter/DeviceAPICharter.html#deliverables), soon to be adopted for all deliverables), and update the test suite accordingly? Even if I see my name in OWNERS, I may not have the cycles to pull off the test suite update at this time, and I'd +1 you to become an owner for the battery status test suite.

On a quick glimpse, I think we need to update the iframe tests and add new tests the Feature Policy integration, and test interesting edge cases such as https://github.com/w3c/battery/pull/13#issuecomment-330586985 

@foolip, this is another API that needs a bunch of manual tests due to its inherent nature, similarly to Vibration API. Is the following guide still considered the state of the art in manual w-p-t testing? :-) 

http://web-platform-tests.org/writing-tests/manual.html

**Comment by Honry** [comment]
@anssiko, according to the [test as you commit](https://w3c.github.io/dap-charter/DeviceAPICharter.html#deliverables) policy, would you please add a label "needs tests" to this PR?

I will add a corresponding PR in w-p-t to reflect the changes and involve my name in OWNERS file. 
Then revisit current test coverage. :)

**Comment by anssiko** [comment]
>would you please add a label "needs tests" to this PR?

Done. Thanks for your help updating this test suite too!

**Comment by Honry** [comment]
>I will add a corresponding PR in w-p-t to reflect the changes and involve my name in OWNERS file.

Done at https://github.com/w3c/web-platform-tests/pull/7944, PTAL.

**Comment by mounirlamouri** [comment]
Sorry for the delay Anssi :( I still disagree with this change as I think it's curing thy symptoms, not the disease. The API should have an implementation that is privacy friendly.

This said, in practice, my main concern is whether this will break web pages. Was a use counter added into Chrome to discover how large the impact of this change would be and whether the deprecation is web-compatible? I don't think we can disable this in Chrome without this information.

**Comment by clelland** [comment]
@mounirlamouri, I don't think we've added a use counter for that into Chrome.

When we were looking at a similar behaviour change for the Fullscreen API, I think we added one that fired every time a same-origin page was blocked from entering fullscreen.

We could add a similar counter here for battery-status, as well as one that fires when a same-origin-as-root-but-embedded-cross-origin page successfully uses the API. If we do it quickly enough, we should be be able to get that change merged for M63, to get some actual data before the end of the year.

**Comment by mounirlamouri** [comment]
I think something that checks the usage in x-origin iframe and another for same origin should be enough.

**Comment by clelland** [comment]
So Chrome doesn't currently restrict `getBattery` at all. Fully complying with this change would block the API in Chrome:

  - in all insecure contexts
  - in any frame which is at a different origin than its parent  (without an appropriate feature policy)
  - in any frame which is at a different origin than the top-level document (again, without an appropriate feature policy)

If we add counters for those situations, then we should have a good idea how much usage would be broken by adopting this change (and also adopting the current spec as-is).

(Also, this is Chrome-specific, but both [caniuse](http://caniuse.com/#search=getBattery) and the [API confluence dashboard](https://web-confluence.appspot.com/#!/catalog?releaseKeys=%5B%22Safari_11.0_OSX_10.12.6%22,%22Chrome_61.0.3163.79_Windows_10.0%22,%22Firefox_55.0_Windows_10.0%22,%22Edge_15.15063_Windows_10.0%22%5D&releaseOptions=null&numAvailable=null&currentPage=0&gap=5&itemPerPage=15&searchKey=%22getbattery%22) suggest that the API is not implemented anywhere else -- if that's wrong, it could change the 'web-compatible' calculation quite a bit)

**Comment by anssiko** [comment]
@clelland, could you help us and add the use counters as you suggested? Getting them into M63 would be great. Thanks!

**Comment by mounirlamouri** [comment]
@clelland Regarding other implementations. Firefox had one but it was disabled when they disabled all the Firefox OS features with the argument that it was leaking too much data. Arguably, they could have reduced the granularity of the shared data.

**Comment by clelland** [comment]
@anssiko -- I've just landed the counters into Chromium now, but it'll take a couple of days to get merge approval for M63.

**Comment by anssiko** [comment]
@clelland, thanks!

I suggest we leave this PR open until we have actual data to make an informed decision, and revisit by the end of this year, or early next year.

**Comment by foolip** [comment]
> @foolip, this is another API that needs a bunch of manual tests due to its inherent nature, similarly to Vibration API. Is the following guide still considered the state of the art in manual w-p-t testing? :-)

http://web-platform-tests.org/writing-tests/manual.html is the documentation for manual tests, but as of today, http://web-platform-tests.org/writing-tests/testdriver.html is the state of the art.

For all of your automation needs, I would run the following steps:
1. If the size and shape of a WebDriver API for automation would be simple, define a WebDriver extension in your spec, or put stuff in the WebDriver spec itself.
1. Wrap that in `test_driver.something()`
1. If that would be too complicated, escalate to public-test-infra and investigate doing a testing API in the style of https://wicg.github.io/webusb/test/

**Comment by anssiko** [comment]
Now we have the Battery Status API use counter stats from Chrome since 25 Oct 2017:

* [Same-origin](https://www.chromestatus.com/metrics/feature/timeline/popularity/2201)
* [Cross-origin](https://www.chromestatus.com/metrics/feature/timeline/popularity/2200)
* [Insecure origin](https://www.chromestatus.com/metrics/feature/timeline/popularity/2199)

Cross-origin usage has been >1.1% and insecure origin usage >0.7% of page loads over the last month.

@mounirlamouri what are your thoughts? A significant number of page loads would be affected by this change.

**Comment by mounirlamouri** [comment]
So, looking at the numbers, results from `UseCounter::CountIfFeatureWouldBeBlockedByFeaturePolicy` is that 90% of calls would be blocked. Also, 60% of calls are made on an insecure origin.

It sounds that the restrictions would virtually shut down the API usage. Now, I'm not fully against as maybe it is being abused but we should be careful and analyse the impact. Ideally, it would be great to have an idea of which websites are using the API but there is no good tools for this :(

**Comment by anssiko** [comment]
@mounirlamouri any precedence on how you've handled similar situations in the past? I assume there are occurrences of APIs being restricted to secure origin.

Indeed without knowing which websites are using the feature in ways that'd be blocked, we don't know whether they'd be able to migrate over to HTTPS and/or add the required Feature Policy tokens to the embedder page. My hunch is requiring FP would help weed out possible malicious usage.

Adding these restrictions might help other implementers to move forward with the feature, so I suggest we analyze this more and seek feedback from other browser vendors too.

Would logging warnings to console on insecure or cross-origin usage help us gather data on the nature of usage in Chrome? I'd assume impacted sites having legitimate usage would let us know if we'd point them to this issue in the warning message.

Any further help and ideas appreciated :-)

**Comment by mounirlamouri** [comment]
I'm not convinced that making the API HTTPS-only is a goal worth achieving. This said, given the current trend, I can see this happening and a deprecation warning would be appropriate like we are doing for Device Orientation.

For FP, what would be the plan? Would the API be denied by default or allowed? Do you know what's the idea for sensors?

**Comment by anssiko** [comment]
>For FP, what would be the plan? Would the API be denied by default or allowed?

This PR proposes that the default allowlist for the Battery Status API is `["self"]` and that the battery promise will be rejected with a `SecurityError` if not allowed to use the policy-controlled feature.

>Do you know what's the idea for sensors?

Concrete sensors that extend `Sensor` throw at construction if they're not allowed to use the policy-controlled feature, see https://w3c.github.io/sensors/#construct-sensor-object

The default allowlist for sensors is similarly `["self"]`, see https://w3c.github.io/sensors/#feature-policy-api

**Comment by clelland** [comment]
> > For FP, what would be the plan? Would the API be denied by default or allowed?

> This PR proposes that the default allowlist for the Battery Status API is `["self"]` and that the battery promise will be rejected with a `SecurityError` if not allowed to use the policy-controlled feature.

That matches my understanding -- so all usage would be allowed on the top-level document, and that document could allow access in cross-origin frames by embedding them with `<iframe allow="battery"></iframe>`, or with an HTTP header that allows it in that origin. Other usage would be denied.

**Comment by RByers** [comment]
>  Ideally, it would be great to have an idea of which websites are using the API but there is no good tools for this :(

There are [some tools](https://www.chromium.org/blink/platform-predictability/compat-tools). After the UseCounters have been in stable for a bit, we could [search HTTP Archive](https://groups.google.com/a/chromium.org/forum/#!topic/blink-api-owners-discuss/uxwEuxCRfGA) for their usage.  Also it's now possible for a Googler to use [CrUX data](https://developers.google.com/web/tools/chrome-user-experience-report/) to get an even clearer view of usage in the wild.

**Comment by RByers** [comment]
FWIW, my hunch is that the [severity of breakage](https://docs.google.com/document/d/1RC-pBBvsazYfCNNUSkPqAVpSpNJ96U8trhNkfV0v9fk/edit#heading=h.u5ya6jvru7dl) is likely to be very small.  If there's a compelling reason to lock this down, then (as long as we have the FP opt-in) I suspect we could get away with it.

All that said, I have heard of people using coarse grained battery status data to do A/B tests for power-saving optimizations.  If that's what a lot of this usage is actually doing in practice (eg. some ad network script), then it seems like a pretty beneficial use case and argument for NOT locking this down (at least until there's a good WebPerf API around power usage).

**Comment by anssiko** [comment]
Thanks @RByers!

I'm hearing we want to wait until we have more UseCounters data to be able to do a search on a bigger data set.

If everyone is happy with that plan, I suggest we park this issue for couple more months and revisit.

File: index.html [hunk 1]
```diff
@@ -135,8 +135,9 @@ <h2>
         Terminology
       </h2>
       <p>
-        The following concepts, terms, and interfaces are defined in
-        [[!HTML5]], [[!ECMASCRIPT]], [[!WEBIDL]], and [[!SECURE-CONTEXTS]]:
+        The following concepts, terms, and interfaces are defined in [[!HTML]],
+        [[!HTML5]], [[!ECMASCRIPT]], [[!WEBIDL]], [[!SECURE-CONTEXTS]], and
```
Comment (TimothyGu) [inline index.html]: What's HTML5 still used for? Everything used should come out of WHATWG's version of the spec.

File: index.html [hunk 2]
```diff
@@ -483,6 +515,20 @@ <h2>
...
     </section>
+    <section>
+      <h2>
+        Feature Policy integration
+      </h2>
+      <p data-link-for="Navigator">
+        The Battery Status API is a <a>policy-controlled feature</a>, as
+        defined by Feature Policy [[!FEATURE-POLICY]]. The <a>feature name</a>
+        for the Battery Status API is "<code>battery</code>". The <a>default
+        allowlist</a> for the Battery Status API is <code>["self"]</code>. When
```
Comment (TimothyGu) [inline index.html]: The allowlist is defined as an [ordered set](https://infra.spec.whatwg.org/#ordered-set), which is a [list](https://infra.spec.whatwg.org/#list). So instead of `["self"]`, « "`self`" ».

File: index.html [hunk 3]
```diff
@@ -255,17 +286,18 @@ <h2>
...
         steps.
         </li>
-        <li>If the <a>browsing context</a> is not a <a>top-level browsing
-        context</a>, then reject this <a>Navigator</a> object's <a>battery
-        promise</a> with a "<a>SecurityError</a>" <a>DOMException</a>, return
-        this <a>Navigator</a> object's <a>battery promise</a> and abort these
-        steps.
+        <li>If the <a>top-level browsing context</a>'s <a>active document</a>'s
+        <a>origin</a> and the <a>origin</a> specified by the <a>current
+        settings object</a> is not <a>same-origin domain</a>, then reject this
```
Comment (TimothyGu) [inline index.html]: same origin-domain

File: index.html [hunk 4]
```diff
@@ -204,6 +205,36 @@ <h2>
...
+        </li>
+        <li>
+          <a href=
+          "https://html.spec.whatwg.org/multipage/origin.html#concept-origin"><dfn>
+          origin</dfn></a>
+        </li>
+        <li>
+          <a href=
+          "https://html.spec.whatwg.org/multipage/origin.html#same-origin-domain">
+          <dfn>same-origin domain</dfn></a>
```
Comment (TimothyGu) [inline index.html]: same origin-domain

File: index.html [hunk 5]
```diff
@@ -255,17 +286,18 @@ <h2>
...
         promise</a> with a "<a>SecurityError</a>" <a>DOMException</a>, return
         this <a>Navigator</a> object's <a>battery promise</a> and abort these
         steps.
         </li>
-        <li>If the <a>browsing context</a> is not a <a>top-level browsing
-        context</a>, then reject this <a>Navigator</a> object's <a>battery
-        promise</a> with a "<a>SecurityError</a>" <a>DOMException</a>, return
-        this <a>Navigator</a> object's <a>battery promise</a> and abort these
-        steps.
+        <li>If the <a>top-level browsing context</a>'s <a>active document</a>'s
```
Comment (TimothyGu) [inline index.html]: https://github.com/w3c/battery/issues/19 doesn't seem to be fixed here.

> Probably you want, hmm, this Navigator object's relevant global object's associated Document's browsing context?

File: index.html [hunk 6]
```diff
@@ -255,17 +286,18 @@ <h2>
         the following steps:
       </p>
       <ol>
-        <li>If the <a>incumbent settings object</a> is not a <a>secure
+        <li>If the <a>current settings object</a> is not a <a>secure
```
Comment (TimothyGu) [inline index.html]: Things in the web platform generally use "relevant settings object of this `Navigator` object." We generally don't want to use "current". See https://html.spec.whatwg.org/multipage/webappapis.html#realms-settings-objects-global-objects, the example starting with "One reason why the relevant concept is generally a better default choice than the current concept"

**Comment by mounirlamouri** [comment]
@RByers I don't know about the severity of breakage because it will mostly depend on how the scripts will handle a failing attempts. Though, I'm more worried about shutting down valid use cases.

**Comment by anssiko** [comment]
@TimothyGu, thanks a lot for your comments! I updated the PR, PTAL.

**Comment by anssiko** [comment]
(Here's an [HTML diff](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fw3c.github.io%2Fbattery%2F&doc2=https%3A%2F%2Frawgit.com%2Fw3c%2Fbattery%2Fissue-10-secure-top-level%2Findex.html) comparing [this branch](https://rawgit.com/w3c/battery/issue-10-secure-top-level/index.html) with [gh-pages](https://w3c.github.io/battery/) to ease review, e.g. at F2F. I enabled pr-preview to automate this for any subsequent PRs.)

File: index.html [hunk 7]
```diff
@@ -255,17 +290,20 @@ <h2>
...
+        <a>Navigator</a> object's <a>battery promise</a> with a
+        "<a>SecurityError</a>" <a>DOMException</a>, return this
+        <a>Navigator</a> object's <a>battery promise</a> and abort these steps.
         </li>
-        <li>If the <a>browsing context</a> is not a <a>top-level browsing
-        context</a>, then reject this <a>Navigator</a> object's <a>battery
-        promise</a> with a "<a>SecurityError</a>" <a>DOMException</a>, return
-        this <a>Navigator</a> object's <a>battery promise</a> and abort these
-        steps.
+        <li>If this <a>Navigator</a> object's <a>relevant global object</a>'s
```
Comment (clelland) [inline index.html:321]: With feature policy integration, I think you can replace this entire paragraph with something like

> If this Navigator object's relevant global object's associated Document is not [allowed to use](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#allowed-to-use) the <code>battery</code> feature, then reject...

Because the default allowlist is `'self'`, that will automatically take care of the same-origin embed case, while allowing cross-origin usage only if explicitly enabled by the embedding document (which this paragraph still prohibits, I think)

File: index.html [hunk 8]
```diff
@@ -483,6 +521,20 @@ <h2>
...
     </section>
+    <section>
+      <h2>
+        Feature Policy integration
+      </h2>
+      <p data-link-for="Navigator">
+        The Battery Status API is a <a>policy-controlled feature</a>, as
+        defined by Feature Policy [[!FEATURE-POLICY]]. The <a>feature name</a>
+        for the Battery Status API is "<code>battery</code>". The <a>default
+        allowlist</a> for the Battery Status API is <code>« "self" »</code>.
```
Comment (clelland) [inline index.html]: Minor nit: the value for the default allowlist would be `'self'`, with single quotes included

You can just say now

> The Battery Status API is a policy-controlled-feature identified by the string "`battery`". It's default allowlist is `'self'`.

File: index.html (hunk 7 above)
Comment (anssiko) [reply index.html:321]: I pushed an update in an attempt to make use of "allowed to use". Please take a look. (I'm happy to see "allowed to use" abstracted out and reusable.)

File: index.html (hunk 8 above)
Comment (anssiko) [reply index.html]: Ditto, pushed an update per your proposal.

File: index.html [hunk 9]
```diff
@@ -297,13 +301,18 @@ <h2>
...
+        the <code>battery</code> feature, then reject this <a>Navigator</a>
+        object's <a>battery promise</a> with a "<a>SecurityError</a>"
+        <a>DOMException</a>, return this <a>Navigator</a> object's <a>battery
+        promise</a> and abort these steps.
+          <div class="note">
+            In other words, this step rejects if the <a>associated
+            <code>Document</code></a>'s <a>browsing context</a>'s <a>active
+            document</a>'s <a>origin</a> is not <a>same origin-domain</a> with
+            the <a>origin</a> of the <a>current settings object</a> of this
+            <a>Navigator</a> object.
```
Comment (clelland) [inline index.html]: You could add something like "unless specifically allowed by the document's feature policy." to this, if you wanted to be clear that it's *possible* to grant access to cross-origin subframes, but you have to be deliberate about it. The default situation is exactly as you describe here.

**Review by clelland — COMMENTED** [review]
One small comment about the non-normative explanation; this looks great otherwise.

**Comment by anssiko** [comment]
@clelland, thanks for your comments! I integrated your suggestion and pushed an update. At F2F, we can gauge whether we have consensus to land this and update the implementation(s) accordingly. I invited @mounirlamouri to join as well.

File: index.html [hunk 10]
```diff
@@ -483,6 +531,18 @@ <h2>
         </table>
       </section>
     </section>
+    <section>
+      <h2>
+        Feature Policy integration
+      </h2>
+      <p data-link-for="Navigator">
+        The Battery Status API is a <a>policy-controlled feature</a> identified
+        by the string "<code>battery</code>". It's default allowlist is
```
Comment (anssiko) [inline index.html:557]: (Note to self: s/It’s/Its/)

**Comment by anssiko** [comment]
F2F Resolution: https://www.w3.org/2018/10/22-dap-minutes.html#x05

**Comment by anssiko** [comment]
(Rebased with gh-pages and fixed merge conflicts in a9e3c94.)

**Comment by anssiko** [comment]
Time for the annual update to this issue ☺️

There are the following [battery use counters](https://cs.chromium.org/chromium/src/third_party/blink/public/mojom/web_feature/web_feature.mojom?type=cs&q=BatteryStatusGetBattery&g=0&l=1627) invoked in [getBattery](https://cs.chromium.org/chromium/src/third_party/blink/renderer/modules/battery/navigator_battery.cc?type=cs&g=0&l=32).

In percentages of Chrome page loads (across all channels and platforms) that use the corresponding feature at least once, as of today:

* 18.474067% [BatteryStatusGetBattery](https://www.chromestatus.com/metrics/feature/popularity#BatteryStatusGetBattery)
* 4.141155% [BatteryStatusCrossOrigin](https://www.chromestatus.com/metrics/feature/popularity#BatteryStatusCrossOrigin)
* 1.652762% [BatteryStatusInsecureOrigin](https://www.chromestatus.com/metrics/feature/popularity#BatteryStatusInsecureOrigin)
* 0.004663% [BatteryStatusSameOriginABA](https://www.chromestatus.com/metrics/feature/popularity#BatteryStatusSameOriginABA)

**Comment by anssiko** [comment]
Merging this PR per F2F resolution https://github.com/w3c/battery/pull/13#issuecomment-533378170. This branch includes https://github.com/w3c/battery/pull/13/commits/3e089152068c61ae97751af8be1bafaaeecbd768 that was part of the resolution.

**Comment by anssiko** [comment]
Thanks to @honry, now also Feature Policy section is updated. Consider #24 to be a part of this PR.

**Comment by anssiko** [comment]
- [x] web-platform-tests https://github.com/web-platform-tests/wpt/pull/7944 
- [ ] Chromium https://crbug.com/1007264
